pkg/server/preprocessor/auth.go

```diff
@@ -68,15 +68,23 @@ func (p AccessKeyValidationPreprocessor) Preprocess(payload *router.Payload, res
 
 // UserAuthenticator provides preprocess method to authenicate a user
 // with access token or non-login user without api key.
+// if ByPassUnauthorized is true, UserAuthenticator will only inject AuthInfoID
+// and related context, and always return StatusOK
+// it is for plugin so even handler or lambda are not user_required,
+// we can still get the user id from context
 type UserAuthenticator struct {
-	ClientKey  string
-	MasterKey  string
-	AppName    string
-	TokenStore authtoken.Store
+	ClientKey            string
+	MasterKey            string
+	AppName              string
+	TokenStore           authtoken.Store
+	ByPassUnauthorized   bool
 }
 
 func (p *UserAuthenticator) Preprocess(payload *router.Payload, response *router.Response) int {
 	if err := checkRequestAccessKey(payload, p.ClientKey, p.MasterKey); err != nil {
+		if p.ByPassUnauthorized {
```

Comment:
Pro tip: you can do it like this:

```
 func (p *UserAuthenticator) Preprocess(payload *router.Payload, response *router.Response) (status int) {
    if (ByPassUnauthorized) {
      defer func() {
        if status == http.StatusUnauthorized {
          status == http.StatusOK;
        }
      }();
    }
```

In other words,

1. Give a parameter name for the return value.
2. If Bypass is true, ask golang to run your function when this function returns (`defer`)
3. In the function, if the return value is unauthorized, change the return value back to OK.

But your code is good as is. So don’t change it now, but it is good to know.

Reply:
Thanks!! It is good to know I could do it this way! Actually at the beginning I used a variable to hold the return code, something like 
```
unauthorizedReturnCode = http.StatusUnauthorized
if (BypassUnauthorized) {
      unauthorizedReturnCode = http.StatusOK;
}
```
return `unauthorizedReturnCode` in the unauthorized case. But I found we also need to reset `response.Err`.
If I do it by `defer` function, then I could reset `response.Err` and `status`. :)